pilot/pkg/xds/xds_test.go

```diff
@@ -543,11 +544,14 @@ spec:
 					fakeOpts.MeshConfig = &meshConfig
 					s := NewFakeDiscoveryServer(t, fakeOpts)
 					for clusterID := range want {
-						p := &model.Proxy{Metadata: &model.NodeMetadata{ClusterID: clusterID}}
-						eps := xdstest.ExtractLoadAssignments(s.Endpoints(p))[tt.serviceCluster]
-						if want := want[clusterID]; !listEqualUnordered(eps, want) {
-							t.Errorf("got %v but want %v for %s", eps, want, clusterID)
-						}
+						retry.UntilSuccess(func() error {
```

Comment:
I am concerned by this change... does this mean we had some guarantees in the old code that are not present in the new code?

Biggest concern is this impacts prod users. Secondary impact is this impacts ALL our unit tests, and we are just fixing a single one.

Reply:
Also need to use UntilSuccessOrFail otherwise this doesn't test anything.

Reply:
In theory, original all tests should wait for some time. before got the expectations. Because the handlers processings are asych.  What makes the test failure i think is the async eds update. 

It does not influence users much.

Reply:
If that was the case I would expect this test (and NDS, etc) to fail before this PR. As is, it passes 100% of the time without this PR and fails regularly with this PR. That leads me to believe the behavior has changed.

Reply:
Chnaged

Reply:
A bunch of other tests fail now with this PR, such as TestServiceScoping/STATIC. I suspect this code: https://github.com/istio/istio/blob/1c82a227a4e258b0db9bb2922484dbd277fd79ac/pilot/pkg/xds/fake.go#L315-L319 is now broken.

Reply:
where did it fail?

Reply:
>If that was the case I would expect this test (and NDS, etc) to fail before this PR

I figured it out, the tests ` NewConfigGenTest ` use the sync memory controller, `memory.NewSyncController`.